Start building Clang runtimes on-demand

chandlerc opened the pull request:
This is the first step to having Clang's runtime libraries fully
available for the Carbon toolchain. This PR focuses on the lowest level
runtimes, the CRT files and the builtins library.

The goal is to intercept Clang runs where it needs these
target-dependent pieces to be available, and build them on demand using
our Clang-running infrastructure. This avoids most of the subprocess
overhead, but there is still some due to missing features in Clang.

This requires exporting the sources for these runtimes from the Bazel
build, and installing them in our target-independent resource directory.
We then build a simplified "build" of these sources within the
`ClangRunner` itself to produce the specific artifacts and layout
expected by Clang.

It also required fixing our use of Clang on macOS to have a default
system root in order to successfully compile or link.

It also required cleaning up how the `ClangRunner` used target
information more generally -- instead of taking the target as
a constructor parameter, it manages its target internally and relies on
the Clang target-specifying command line flags.

I looked at whether we could split this into another layer separate from
the `ClangRunner`, but that proved frustratingly difficult to manage.
While we support building these on-demand as part of a detected link,
that doesn't seem feasible as we don't have the necessary separation
between compilation runs of Clang and link runs of Clang. However,
I have tried to factor the internals to provide as clear of separation
as I could across these.

I have also created a stand-alone subcommand to directly build the
runtimes which allows for easy testing. It also supports building them
into a specific directory, and that directory can in turn be passed to
a Clang invocation. This is designed to work both at the API level with
`ClangRunner` and at the subcommand level.

Currently, the only part of the commandline that is detected and
forwarded to the runtimes build is the target. Eventually, the plan is
to expand this so that we can build a maximally tailored set of runtimes
for a given compilation.

The other big TODO here is to actually implement caching storage of
these runtimes so they aren't built on every execution. Right now, this
uses a somewhat hack-y build of a temporary directory, but this isn't
expected to be suitable long-term. Building these runtimes on *every*
link makes those commands take approximately 15 seconds with an ASan
build like our default development build, and just over 2 seconds in an
optimized build. Because of this, I've kept all of this disabled by
default for now. The goal is that once caching and some other
improvements land, we can enable this by default.

chandlerc committed:
Start building Clang runtimes runtimes on-demand

Note: this is disabled by default initially, but can be enabled with the
`--build-runtimes` to the `clang` subcommand.

This is the first step to having Clang's runtime libraries fully
available for the Carbon toolchain. This PR focuses on the lowest level
runtimes, the CRT files and the builtins library.

The goal is to intercept Clang runs where it needs these
target-dependent pieces to be available, and build them on demand using
our Clang-running infrastructure. This avoids most of the subprocess
overhead, but there is still some due to missing features in Clang.

This requires exporting the sources for these runtimes from the Bazel
build, and installing them in our target-independent resource directory.
We then build a simplified "build" of these sources within the
`ClangRunner` itself to produce the specific artifacts and layout
expected by Clang.

It also required fixing our use of Clang on macOS to have a default
system root in order to successfully compile or link.

It also required cleaning up how the `ClangRunner` used target
information more generally -- instead of taking the target as
a constructor parameter, it manages its target internally and relies on
the Clang target-specifying command line flags.

I looked at whether we could split this into another layer separate from
the `ClangRunner`, but that proved frustratingly difficult to manage.
While we support building these on-demand as part of a detected link,
that doesn't seem feasible as we don't have the necessary separation
between compilation runs of Clang and link runs of Clang. However,
I have tried to factor the internals to provide as clear of separation
as I could across these.

I have also created a stand-alone subcommand to directly build the
runtimes which allows for easy testing. It also supports building them
into a specific directory, and that directory can in turn be passed to
a Clang invocation. This is designed to work both at the API level with
`ClangRunner` and at the subcommand level.

Currently, the only part of the commandline that is detected and
forwarded to the runtimes build is the target. Eventually, the plan is
to expand this so that we can build a maximally tailored set of runtimes
for a given compilation.

The other big TODO here is to actually implement caching storage of
these runtimes so they aren't built on every execution. Right now, this
uses a somewhat hack-y build of a temporary directory, but this isn't
expected to be suitable long-term. Building these runtimes on *every*
link makes those commands take approximately 15 seconds with an ASan
build like our default development build, and just over 2 seconds in an
optimized build. Because of this, I've kept all of this disabled by
default for now. The goal is that once caching and some other
improvements land, we can enable this by default.

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

jonmeow commented on bazel/llvm_project/0004_Add_support_for_custom_rules_to_build_builtins.patch:
```diff
@@ -0,0 +1,26 @@
++
++# Allow building custom rules for runtimes.
++exports_files(glob(["**"]))
++
```
Had you considered consolidating patches with `0002_Added_Bazel_build_for_compiler_rt_fuzzer.patch`? e.g. by making that a local build rule and using your new export to underly it; upstreaming patches; or by just concatenating the patch.

My concern is that, if we have a patch conflict, the cost of fixing conflicts is probably going to be per-patch file.

chandlerc replied:
It seemed simpler to keep these broken out and orthogonal until we've somewhat settled on what we want here... We've already gone through a few iterations with builtins, and I still need to get libunwind and libc++ working.

My preference would be to get these things working, and then look across them and see what makes sense to upstream to LLVM. And if we can't upstream them, how to organize them as patches, although I'm somewhat optimistic that once we have this reasonably sorted out we'll have a good case for upstreaming.

Does that seem like a reasonable approach?

jonmeow replied:
Glancing at history, this file has only broken once, so okay. Would be appreciated if upstreaming occurs, particularly to avoid maintenance cost.

chandlerc replied:
FWIW, now that this is a much more rich build setup, it seems even more worthwhile to try to upstream...

jonmeow commented on toolchain/base/runtime_sources.bzl:
```diff
@@ -0,0 +1,435 @@
+        BUILTINS_I386_SRCS,
+    ]
+    for src in srcs
+]
```
Why the individual lists of source files for this with exports_files, versus filegroups and globs?

To note, comparing with https://github.com/llvm/llvm-project/blame/6d03f51f0c59171f1ec3c5cc5c1fe71c30956273/compiler-rt/lib/builtins/CMakeLists.txt and the file's history in full, it seems like files are added/removed on a not-quite monthly basis.

chandlerc replied:
I didn't see any reasonably clear globbing structure. The separate lists try to directly follow the structure of the upstream CMake files.

As for why a list here rather than a filegroup -- we need to generate source code to describe these files as well, and this seemed like a much easier way to manage than trying to extract the right information from a filegroup.

jonmeow replied:
> I didn't see any reasonably clear globbing structure. The separate lists try to directly follow the structure of the upstream CMake files.

I see some of these lists require a lot of hand-crafting, but maybe we could use globs for other cases?

e.g., giving a few examples (particularly for the "generic" since it seems like the most challenging to get right):

```
builtins_bf16_patterns = ["extendbfsf2.c", "*bf2.c"]
builtins_macos_patterns = ["atomic_*"]

# These have some globs that could be used to elide a few files
# (e.g., extend*tf2.c), but assume copying the full lists for the moment.
builtins_tf_srcs = [ ... ]
builtins_x86_fp80_srcs = [ ... ]

filegroup(
    name = "builtins_generic",
    srcs = glob(
        ["*", "cpu_model/cpu_model.h"],
        exclude =
            ["*.txt", "clear_cache.c", "crt*", "extendhfdf2.c", "emutls.c", "enable_execute_stack.c", "eprintf.c"] +
            builtins_bf16_patterns + builtins_macos_patterns + builtins_tf_srcs + builtins_x86_fp80_srcs,
    )
)
filegroup(
    name = builtins_bf16,
    srcs = glob(builtins_bf16_patterns)
)
filegroup(
    name = builtins_macos,
    srcs = glob(builtins_macos_patterns)
)
filegroup(
    name = builtins_x86_64,
    srcs = glob(["x86_64/*"], exclude = ["**/chkstk.S"])
)
filegroup(
    name = builtins_i386,
    srcs = glob(["i386/*"], exclude = ["**/chkstk.S"])
)
```

Glancing at some commits, to try to compare with how this would work versus breakages:

- https://github.com/llvm/llvm-project/commit/6d03f51f0c59171f1ec3c5cc5c1fe71c30956273 adding `extendhfdf2.c`: the glob would cause this to be added to builtins_generics. You've excluded that: is the current PR correct and adding it would be wrong, or would the glob behavior be correct? Noting the chance of the former, I've asked for a comment in the BUILD since it's canonically in GENERIC_SOURCES.
- https://github.com/llvm/llvm-project/commit/cb5d866feea72e0a846b4e7b921aaf6e70e196f1 adding `arm_apple_sme_abi.s` -- I'm assuming this would be manually excluded from the aarch64 glob, so I'd call that a break for glob only.
- https://github.com/llvm/llvm-project/commit/28ae2ff2a44c0a5d09ce5e88d8d8f3309b6f127f adding `truncxfhf2.c` -- added to x86_80, which would be a break either way.
- https://github.com/llvm/llvm-project/commit/cb4f4a8a4dd18bf00604b49faadd7b0ee4394d3d added and removed aarch64 files -- looks like a break for the per-file list only.
- https://github.com/llvm/llvm-project/commit/bbd99d903ec956b5cc87c544854337da8db2d2bf removed an aarch64 file -- I'm assuming it would've been included, so a break for a per-file list only.
- https://github.com/llvm/llvm-project/commit/a5dd6463608bb09404e8a898ed337ef0b4a292c4 added a file to x86_80, so similar to above would be a break either way.
- https://github.com/llvm/llvm-project/commit/28e4aad45a64ec893c02f21b9c2afe7efe5f4a2a added a file to BF16_SOURCES, so a break for a per-file list only.

To summarize, looking at commits since November:

- 1 commit that might have been overlooked for the current PR; depending on the answer, it would only break globs, or per-file lists.
- 2 commits that would break either way.
- 1 commit that would only break a glob approach.
- 3 commits that would only break a per-file list approach.

Depending on which way https://github.com/llvm/llvm-project/commit/6d03f51f0c59171f1ec3c5cc5c1fe71c30956273 falls, maybe globs would avoid enough breaks that they'd be worth it?

> As for why a list here rather than a filegroup -- we need to generate source code to describe these files as well, and this seemed like a much easier way to manage than trying to extract the right information from a filegroup.

[manifest code](https://github.com/carbon-language/carbon-lang/blob/trunk/bazel/manifest/defs.bzl) might be useful as an example for doing this (could also refactor a little to share `_get_files`)

chandlerc replied:
Ok, this has been completely reworked to use filegroups, globs, and successively excluding prior glob patterns to build up the structure without an explicit list.

In good news, it even caught another bug: `clear_cache.c` should have been included but I missed that one when reading the CMake code.

I also ended up being able to document the specific file name structure that most of the builtins are following, so it ends up pretty clean and providing a nice structural map of the files here. All of this is now in the patch to compiler-rt.

Here, I also refactored this a bunch to try and get a better model where we only need a single dict of the different file groups, and then use that to build up everything else. Hope that helps some with reducing the repetition and boiler plate.

jonmeow commented on toolchain/base/runtime_sources.bzl:
```diff
@@ -0,0 +1,435 @@
+#    "x86_64-unknown-linux-gnu": BUILTINS_GENERIC_SRCS + BUILTINS_BF16_SRCS + BUILTINS_TF_SRCS + BUILTINS_X86_ARCH_SRCS + BUILTINS_X86_FP80_SRCS + [
+#    "i386-unknown-linux-gnu": BUILTINS_GENERIC_SRCS + BUILTINS_BF16_SRCS + BUILTINS_X86_ARCH_SRCS + BUILTINS_X86_FP80_SRCS + [
+
+#[carbon_cc_runtime_static_library(
```
What's the purpose of this?

chandlerc replied:
Sorry, this was left over. It was a remnant of a previous iteration I was keeping around until I had it fully re-implemneted in this version, but that's done now. Removed.

jonmeow commented on toolchain/base/runtime_sources.bzl:
```diff
@@ -0,0 +1,435 @@
+    h_file = ctx.actions.declare_file(ctx.label.name)
+    ctx.actions.write(h_file, _TEMPLATE.format(
+        crtbegin = '"builtins/' + BUILTINS_CRTBEGIN + '"',
+        crtend = '"builtins/' + BUILTINS_CRTEND + '"',
```
Similar to the above:
```suggestion
        crtbegin = '"builtins/{0}"'.format(BUILTINS_CRTBEGIN),
        crtend = '"builtins/{0}"'.format(BUILTINS_CRTEND),
```

jonmeow commented on toolchain/base/runtime_sources.bzl:
```diff
@@ -0,0 +1,435 @@
+"""
+
+def _make_srcs_strings(dir, srcs):
+    return "\n".join(['    "' + dir + "/" + src + '",' for src in srcs])
```
Perhaps `.format`?
```suggestion
    return "\n".join(['    "{0}/{1}",'.format(dir, src) for src in srcs])
```

jonmeow commented on toolchain/base/runtime_sources.bzl:
```diff
@@ -0,0 +1,435 @@
+        x86_fp80_srcs = _make_srcs_strings("builtins", BUILTINS_X86_FP80_SRCS),
+        aarch64_srcs = _make_srcs_strings("builtins", BUILTINS_AARCH64_SRCS),
+        x86_64_srcs = _make_srcs_strings("builtins", BUILTINS_X86_64_SRCS),
+        i386_srcs = _make_srcs_strings("builtins", BUILTINS_I386_SRCS),
```
Why is `builtins` a parameter if it's always passed the same? Had you considered making the file format a constant string `'"builtins/%s"'`, also for crtbegin/crtend?

For example, _make_srcs_strings could also be written as (indenting `{i386_srcs}` and the like by 4 spaces in the template to make this work best):

```
_FILE_FORMAT = '"builtins/{0}"'

def _make_srcs_strings(srcs):
    return ",\n   ".join(_FILE_FORMAT.format(src) for src in srcs)
```

chandlerc replied:
I'm expecting to have other runtimes here, including at least the following;

- `crt` for the C *language* runtimes: just the infra to run global constructors & destructors
- `builtins`: the compiler-builtins library, mirroring `libgcc`
- `libunwind`: the unwinding library
- `libc++`
- `libc++abi`

And eventually, we may want:
- All the sanitizer runtimes
- profiling runtimes

Some parts of LLVM consider the CRT bits to be part of the "builtins" library, other parts keep them separate. I've ended up with a bit of a mix and tried to consolidate on a more consistent naming. I've also added a comment to the `.bzl` file to outline the structure here.

Does that all make sense?

jonmeow commented on toolchain/base/runtime_sources.bzl:
```diff
@@ -0,0 +1,435 @@
+    return "\n".join(['    "' + dir + "/" + src + '",' for src in srcs])
+
+def _generate_runtime_sources_h_rule(ctx):
+    h_file = ctx.actions.declare_file(ctx.label.name)
+    ctx.actions.write(h_file, _TEMPLATE.format(
```
Had you considered using `ctx.actions.expand_template`? It'd allow you to move the template file into a file, versus embedding here. Also, it might let you drop the `{{` duplication.

I think `expand_template` and the normal `substitutions` would be a pretty straight drop-in.

I was looking more closely at [computed_substitutions](https://bazel.build/rules/lib/builtins/actions#expand_template), but it'd be more verbose. IIUC, it'd look roughly like:

```
srcs_dict = ctx.actions.template_dict()
srcs_dict.add_joined(
    "{bf16_srcs}",
    depset(BUILTINS_BF16_SRCS),
    map_each = lambda file: file,
    join_with = "\n",
    format_joined = '"builtins/%s",',
)
...
computed_substitutions = srcs_dict,
```

Note, template_dict isn't clearly documented, so I'm guessing based on example:
https://github.com/bazelbuild/rules_java/blob/master/java/bazel/rules/bazel_java_binary.bzl#L234-L242

chandlerc replied:
Similar to one of the other comments, I'm basing this off the `.def` generation for CLI tools. If we'd like to move to more use of another pattern with a template file, I'm game, but maybe that makes sense as a follow-up?

jonmeow replied:
I'm fine separating this out, though note I'm asking separately what you're comparing to.

jonmeow commented on toolchain/base/runtime_sources.bzl:
```diff
@@ -0,0 +1,435 @@
+
+    The two-step process is necessary to avoid trying to compile or otherwise
+    process the generated file as something other than a textual header.
+    """
```
This comment is confusing to me, does it reflect the current implementation of `generate_runtime_sources_h`? What `.def` file is this referring to?

chandlerc replied:
Sorry, this is a stale comment -- I started by copying the `.bzl` file for LLVM command line tools. Updated the comment.

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+  return RunInternal(args, target, std::nullopt);
+}
+
+auto ClangRunner::BuildTargetResourceDir(llvm::StringRef target)
```
105 lines is kind of long for a function. WDYT of breaking it up?

chandlerc replied:
Sure, factored more steps out.

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+  }
+}
+
+static auto ComputeTarget(llvm::ArrayRef<llvm::StringRef> args) -> std::string {
```
Maybe my head's too much in bazel territory, but maybe it's worth commenting what kind of target this is computing? Or perhaps the heuristic used for coomputation?

chandlerc replied:
Added a fairly detailed comment here because there is a lot going on that we should write down... That said, let me know if more is needed.

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+ClangRunner::~ClangRunner() {
+  if (!tmp_dir_.empty()) {
+    std::error_code ec;
+    if (auto number = std::filesystem::remove_all(tmp_dir_, ec); number == 0) {
```
Why the temporary here? It seems unused, so maybe just remove it? If you keep it to be used somewhere, consider a name like "count"; "number" is a little too generic.

```suggestion
    if (std::filesystem::remove_all(tmp_dir_, ec) == 0) {
```

jonmeow commented on toolchain/install/busybox_info.h:
```diff
@@ -35,8 +35,19 @@ struct BusyboxInfo {
+
+    // Unset the override environment variable so that subsequent subprocessing
+    // back into the busybox doesn't pick it up and uses the subprocesses's
+    // argv[0]. This is essential to, for example, allow the `clang` driver to
```
```suggestion
    // back into the busybox doesn't pick it up, and instead uses the subprocess's
    // argv[0]. This is essential to, for example, allow the `clang` driver to
```

Or maybe just remove the second half, since it's restoring the default:

```suggestion
    // back into the busybox doesn't pick it up. This is essential to, for example, allow the `clang` driver to
```

jonmeow commented on toolchain/install/busybox_info.h:
```diff
@@ -35,8 +35,19 @@ struct BusyboxInfo {
+    // Unset the override environment variable so that subsequent subprocessing
+    // back into the busybox doesn't pick it up and uses the subprocesses's
+    // argv[0]. This is essential to, for example, allow the `clang` driver to
+    // subprocess back into a `clang` invocation with an internal CC1
```
What you mean by a "`clang` invocation"? Perhaps:

```suggestion
    // subprocess back into a `clang` busyboxed binary with an internal CC1
```

jonmeow commented on toolchain/driver/clang_runner_test.cpp:
```diff
@@ -17,26 +17,58 @@
+}
+
+class ClangRunnerTest : public ::testing::Test {
+ public:
```
Note for tests, it's typically `protected`: https://google.github.io/styleguide/cppguide.html#Access_Control

```
 protected:
```

chandlerc replied:
We have a specific style rule that suggests `public`, and all current fixtures in the codebase follow that rule: https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/cpp_style_guide.md#:~:text=For%20Access%20Control%2C%20specifically%20for%20test%20fixtures%20in%20.cpp%20files%2C%20we%20use%20public%20instead%20of%20protected.%20This%20is%20motivated%20by%20the%20misc%2Dnon%2Dprivate%2Dmember%2Dvariables%2Din%2Dclasses%20tidy%20check.

jonmeow commented on toolchain/driver/clang_runner.h:
```diff
@@ -62,7 +87,19 @@ class ClangRunner : ToolRunnerBase {
+                    llvm::StringRef out_path) -> void;
+
+  auto BuildBuiltinsFile(llvm::StringRef target, llvm::StringRef src_file,
+                         llvm::StringRef out_path) -> void;
```
Can you add comments about what these do? It's hard for me to guess from function names.

chandlerc replied:
Done.

jonmeow commented on toolchain/driver/clang_runner.h:
```diff
@@ -62,7 +87,19 @@ class ClangRunner : ToolRunnerBase {
+  auto BuildBuiltinsFile(llvm::StringRef target, llvm::StringRef src_file,
+                         llvm::StringRef out_path) -> void;
+
+  std::filesystem::path tmp_dir_;
```
This seems to have a semantic that, if set, the directory will be cleaned up on destruction. Is that worth noting?

chandlerc replied:
Done.

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+    if (arg.size() == 2) {
+      return arg[1] == 'c' || arg[1] == 'E' || arg[1] == 'S';
+    }
+    if (arg == "-fsyntax-only" || arg == "--version" || "--help" || "/?") {
```
Won't `/?` hit the `!= '-'` test? Also note missing comparisons.

```suggestion
    if (arg == "-fsyntax-only" || arg == "--version" || arg == "--help") {
```

chandlerc replied:
Doh!

And really, I think I was overcomplicating all of this. Replaced it with something much less likely to get wrong. Sadly, its somewhat annoying to test this as it mostly just results in things being slow...

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+
+static auto ComputeTarget(llvm::ArrayRef<llvm::StringRef> args) -> std::string {
+  std::string target = llvm::sys::getDefaultTargetTriple();
+  for (auto [i, arg] : llvm::enumerate(args)) {
```
Had you considered a reverse iteration? One possible reason I can think of not to reverse iterate would be if something *looks* like a flag but is an *argument* to a flag, but this doesn't seem to handle that (e.g., `--flag-with-arg -m32`, but maybe that's non-critical?).

Note, because `reverse(enumerate(...))` doesn't work well, you'd need something like:

```
  for (auto [i, arg] : llvm::reverse(llvm::zip(llvm::seq(args.size()), args))) {
```

The main upshot of that is you could directly return when something like a target is found, instead of iterating through all arguments.

chandlerc replied:
Correct that this doesn't handle separate values from the flag. I don't know of a reasonable way to handle that without handing the arguments to Clang itself to parse twice.

Regarding reversing the order, it seems like this would add quite a bit of complexity to implement things like `-m32` and `-m64`? Right now, it can be modeled as a state machine, with those flags mutating whatever our current state is. To work in reverse, we'd have to explicitly track this state as well, and then later on apply the update. The simplest way I've found to handle the last-match-wins semantic has been to walk forwards and mutate as you go... but I'm open to a suggestion for a simplification by going in reverse, just not seeing it myself.

Note that I had a bug here in how the CL driver was handled and I've fixed that.

jonmeow replied:
Looking at the `cl` handling and how it defaults, yeah maybe it's just not worth it.

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+      return true;
+    }
+    // Everything else is rare and likely not worth the cost of searching for
+    // since its fine to have false negatives.
```
```suggestion
    // since it's fine to have false negatives.
```

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -157,4 +399,98 @@ auto ClangRunner::Run(llvm::ArrayRef<llvm::StringRef> args) -> bool {
+// TODO: Replace this with a filesystem cache, and proper management of
+// concurrent cache access. There are a bunch of hacks in this function that
+// aren't realistically reliable for larger use, and only designed as in interim
+// state to separate the implementation of the caching behavior.
```
Minor, but maybe move this to the header for visibility?

chandlerc replied:
Done.

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -157,4 +399,98 @@ auto ClangRunner::Run(llvm::ArrayRef<llvm::StringRef> args) -> bool {
+    name = llvm::formatv("carbon-runtimes.{0:x}.{1}", salt, attempt).str();
+  }
+
+  // WHen we successfully create a temporary directory, commit it to the object.
```
```suggestion
  // When we successfully create a temporary directory, commit it to the object.
```

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+  std::filesystem::path lib_path = resource_dir_path / "lib";
+  std::filesystem::create_directory(lib_path, ec);
+  lib_path /= target.str();
+  std::filesystem::create_directory(lib_path, ec);
```
WDYT of:

```suggestion
  std::filesystem::path lib_path = resource_dir_path / "lib" / target.str();
  std::filesystem::create_directories(lib_path, ec);
```

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+    BuildCrtFile(target, RuntimeSources::CrtEnd, end_o_path.native());
+  }
+
+  llvm::SmallVector<llvm::StringRef, 512> src_files;
```
Why 512 here and on objs? For objs in particular, I kind of wonder of `0` would be better to get a single allocation from `reserve`. But, this too doesn't seem tightly associated with the probably-known number of files: maybe the file count could be exposed as a constant and used here?

chandlerc replied:
There is a reasonable small and finite number of sources (and objects) we're ever going to need here, but not a constant. So I picked a power of two that was large enough. But its not worth arguing about or creating a named constant so I've just removed it.

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+      append_src_files(llvm::ArrayRef(RuntimeSources::BuiltinsI386Srcs));
+    }
+  } else {
+    CARBON_FATAL("Target architecture is not supported: {0}", target);
```
Would it be worth returning an error instead of CHECK-failing on `target` flags?

chandlerc replied:
This is another place where we should diagnose, added a similar TODO.

jonmeow commented on toolchain/base/runtime_sources.bzl:
```diff
@@ -0,0 +1,435 @@
+    The two-step process is necessary to avoid trying to compile or otherwise
+    process the generated file as something other than a textual header.
+    """
+    generate_runtime_sources_h(name = "runtime_sources.h")
+    cc_library(
+        name = name,
+        hdrs = ["runtime_sources.h"],
+        **kwargs
+    )
```
Minor, but had you considered putting these rule invocations in the BUILD? I was thinking it might be helpful for people trying to figure out where the `runtime_sources.h` include comes from to have it with other content.

chandlerc replied:
I'm following the same pattern as the command line tool macro, and would prefer to keep them somewhat consistent... I can make the header file name a parameter to fully match that rule if that would work better? Or maybe there is a follow-up to both of these? I don't have a particularly strong opinion about where the rule lives - all of the options feel a bit awkward.

jonmeow replied:
Can you link to the pattern you're referring to, to help me understand the consistency? I initially thought you meant the [run_tool rule](https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/run_tool.bzl), or maybe the install directory's [symlink rules](https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/install/symlink_helpers.bzl), but you must mean something else?

chandlerc replied:
I was looking at this: https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/base/llvm_tools.bzl

jonmeow replied:
Thanks. FWIW I think `llvm_tools` seems easier to understand in the `BUILD` because it includes the filename as a parameter, rather than hardcoding it. To put this differently, it feels more reusable when it's a parameter, and `bzl` is for reusable or special things beyond the rule invocations of a `BUILD`. Also the filename and target name seem closely tied, so setting them in the same place is nice.

That said, for things that aren't reusable, I lean more in the `run_tool` direction: use `bzl` functions as little as possible, putting more in the `BUILD`. That also would still have the effect of setting both the filename and target name in the same place. If you're okay with that, maybe it's more of a separate PR for cleanup?

i.e., maybe no changes here, but I do think we should align how we're using one-off build rules:

- Invoke one-off rule directly from `BUILD`, as in `run_tool` (and the simpler `pkg_naming_variables`)
- Invoke one-off rule using wrapper function in `bzl`
  - With one name in `BUILD` and one name in `bzl`, as here
  - With both names in `BUILD`, as in `llvm_tools`

jonmeow commented on toolchain/base/runtime_sources.bzl:
```diff
@@ -0,0 +1,435 @@
+
+BUILTINS_HOSTED_SRCS = [
+    #"emutls.c",
+    #"enable_execute_stack.c",
```
Why do these get commented out?

jonmeow commented on toolchain/base/runtime_sources.bzl:
```diff
@@ -0,0 +1,435 @@
+constexpr inline llvm::StringLiteral CrtBegin = {crtbegin};
+constexpr inline llvm::StringLiteral CrtEnd = {crtend};
+
+constexpr inline llvm::StringLiteral BuiltinsGenericSrcs[] = {{
```
Out of curiosity, why `[]` vs `std::array` for these?

chandlerc replied:
It seemed simpler? I try to avoid using `std::array` unless its specifically needed for some reason.

In this case, it also doesn't work well. You can't use `llvm::StringLiteral` with `std::array` without specifying an explicit length. So you either have to specify a length which seems a bit redundant, or you have to use just `std::array` which will end up as an array of pointers without the nice API for the strings themselves.

jonmeow commented on toolchain/base/runtime_sources.bzl:
```diff
@@ -0,0 +1,435 @@
+constexpr inline llvm::StringLiteral CrtBegin = {crtbegin};
+constexpr inline llvm::StringLiteral CrtEnd = {crtend};
+
+constexpr inline llvm::StringLiteral BuiltinsGenericSrcs[] = {{
```
Why "builtins" for these lists, but not BUILTINS_CRTBEGIN/END? Maybe it could be removed from all of them?

chandlerc replied:
I removed `BUILTINS_` from the CRT ones to be more consistent in calling those "CRT" things and these the sources for the builtins library. Sorry for the confusion -- LLVM mixes up the terminology here and that bled over into this code.

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+  // restore the previous setting at the end.
+  auto restore_leak_flag = llvm::make_scope_exit(
+      [&, orig_flag = enable_leaking_] { enable_leaking_ = orig_flag; });
+  enable_leaking_ = false;
```
The comment says "while building the target resource dir", but this also disables during `RunInternal`, doesn't it? Is the comment what's intended, or is the scope of `restore_leak_flag` what's intended?

chandlerc replied:
No, this was an error when I factored out the code into `RunInternal`. Fixed, it should only apply when building the target resource directory.

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+      std::filesystem::create_directory(obj_path, ec);
+      obj_file = obj_file.drop_front(slash_offset + 1);
+    }
+    obj_path.append(obj_file.begin(), obj_file.end());
```
Is this intended to create the parent directory structure? Had you considered something like:

```suggestion
    std::filesystem::path obj_path = src_file;
    std::filesystem::create_directories(obj_path.parent_path(), ec);
```

chandlerc replied:
I seem to have forgotten about `create_directories` entirely and re-implemented.. doh!

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,227 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+      target_triple.isOSDarwin() ? llvm::object::Archive::K_DARWIN
+                                 : llvm::object::Archive::K_GNU,
+      /*Deterministic=*/true, /*Thin=*/false);
+  CARBON_CHECK(!archive_err, "Error writing archive: {0}", archive_err);
```
Should these also have TODOs to diagnose instead of checking, similar to 243?

chandlerc replied:
These aren't user errors, these represent some kind of system failure like having a broken filesystem... I'm not really sure we have a great way of diagnosing this kind of failure mode at all right now, but we should think about how we want that to work I guess.

jonmeow replied:
But for example this could be a full disk (or even just a permissions error), wouldn't it be better to print a diagnostic and error out, than to crash?

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -89,8 +300,31 @@ auto ClangRunner::Run(llvm::ArrayRef<llvm::StringRef> args) -> bool {
+  // TODO: A different system root should be used for iOS, watchOS, tvOS.
+  // Currently, we're only targeting macOS support though.
+  if (target_triple.isMacOSX()) {
+    // This is the default CLT system root, shown by `xcrun --show-sdk-path`.
+    // We hard code it here to avoid the overhead of subprocessing to `xcrun` on
+    // each Clang invocation, but this may need to be updated to search or
+    // reflect macOS versions if this changes in the future.
+    driver.SysRoot = "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk";
```
Just to note, this looks fairly stable. I was searching and found https://github.com/KLayout/klayout/blob/449a9a968e979143f5ef638e0e26ac41e66e8a12/macbuild/build4mac_env.py#L164, which seems to indicate that macos has been using this since 12 (at least)

chandlerc replied:
Yeah, I'm optimistic but wanted to flag how we could revisit it if needed in the comment.

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -157,4 +399,98 @@ auto ClangRunner::Run(llvm::ArrayRef<llvm::StringRef> args) -> bool {
+
+  // Make a temporary directory exclusive to this builder.
+  std::filesystem::path tmp_dir;
+  if (char* tmpdir_env = getenv("TEST_TMPDIR"); tmpdir_env != nullptr) {
+    tmp_dir = std::string(tmpdir_env);
```
Thoughts on using the implicit cast here?

```suggestion
  if (const char* tmpdir_env = getenv("TEST_TMPDIR"); tmpdir_env != nullptr) {
    tmp_dir = tmpdir_env;
```

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -157,4 +399,98 @@ auto ClangRunner::Run(llvm::ArrayRef<llvm::StringRef> args) -> bool {
+  // of a hack perfect, but should be fine until we implement caching at which
+  // point we'll need to do something better here.
+  auto salt = reinterpret_cast<uintptr_t>(this);
+  std::string name = llvm::formatv("carbon-runtimes.{0:x}", salt).str();
```
Why doesn't this use something like `mkdtemp` or `llvm::sys::fs::createUniqueDirectory`?

chandlerc replied:
None were obvious about how to reliably make this work, especially with `TEST_TMPDIR` and potentially parallel running instances...

I thought about both, but ended up spending some time to build somewhat more custom logic as implemented here that has some custom logic compared to the LLVM one. The max attempts thing is directly based on the LLVM function. But then I realized that none of this would even be needed once we have caching and so all of this needed to be thrown away in the caching version.

That said, I also thought about trying to clean this up into something more broadly usable or actually finding a way to re-use the existing things even though it didn't seem easy to do at first... but given that all of this should be replaced with a caching approach, that seemed like wasted work.

So I added all of the TODOs about this code being bad and something we should replace... Not sure how you'd like to move forward here... But my preference is to not spend more time on this code to fit it into `mkdtemp` or the LLVM version (or debugging why those behave differently), and instead work on the caching version.

jonmeow replied:
Given the TODO to remove, okay. I'll note though, my impression of the LLVM API was that it'd approximately be `std::error_code ec = llvm::sys::fs::createUniqueDirectory(tmp_dir / "carbon-runtimes", tmp_dir_);`, replacing lines 367-391. But I'm okay if you're already preparing to work on the TODO to replace this with a filesystem cache.

chandlerc replied:
Yep, just to be clear, the very next thing I'm planning to do is switch to a caching approach.

jonmeow commented on toolchain/driver/clang_runner.h:
```diff
@@ -43,13 +45,36 @@ class ClangRunner : ToolRunnerBase {
+  // resources.
+  //
+  // This method can be used to avoid building target-dependent resources when
+  // unnecessary, but not all Clang command lines will work correct.
```
```suggestion
  // unnecessary, but not all Clang command lines will work correctly.
```

jonmeow commented on toolchain/driver/clang_runner.h:
```diff
@@ -43,13 +45,36 @@ class ClangRunner : ToolRunnerBase {
+  // contains all the target independent files such as headers. However, for
+  // target-specific files like runtimes, we build those on demand here and
+  // return the path.
+  auto BuildTargetResourceDir(llvm::StringRef target) -> std::string;
```
Could this and `RunTargetIndependentCommand` be private, or are they called from elsewhere? If it's just for tests, maybe those could get friend access, so that the public API is limited?

chandlerc replied:
Both seemed meaningful and useful to clients, so I wanted them to be public. For example, I'm not sure how we want the `carbon link` command to eventually interact with buliding the resource directory. I can also see wanting `carbon compile` to potentially reach for the target independent command interface.

It seemed reasonable to me to have a plausible public API surface here without having users today?

jonmeow reviewed (commented):
Nice work getting this done!

chandlerc reviewed (commented):
Update, PTAL.

jonmeow commented on toolchain/base/runtime_sources.bzl:
```diff
@@ -0,0 +1,440 @@
+    "divsi3.c",
+    "divti3.c",
+    "extendsfdf2.c",
+    "extendhfsf2.c",
```
https://github.com/llvm/llvm-project/commit/6d03f51f0c59171f1ec3c5cc5c1fe71c30956273 added extendhfdf2.c, should it be included here? (AFAICT, that commit should be present since it's April 16 and MODULE.bazel says April 17)

chandlerc replied:
Yeah, this was added after I wrote all the compiler-rt code. This PR has taken a long time to get ready.

Regardless, obsoleted by moving all the file logic to globs in the compiler-rt build file.

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,179 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+    return arg == "-c" || arg == "-E" || arg == "-S" ||
+           arg == "-fsyntax-only" || arg == "--version" || arg == "--help" ||
+           arg == "/?";
+    return true;
```
```suggestion
```

chandlerc replied:
:sigh: why did nothing complain about this... sorry.

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,179 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+}
+
+// Searches an argument list to a Clang execution to determine the expected
+// target string, suitable for use with `llvm::Triple`, should be used.
```
Small grammar nit, maybe one of:
```suggestion
// target string, suitable for use with `llvm::Triple`, that should be
// used.
```
```suggestion
// target string, suitable for use with `llvm::Triple`.
```

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +50,179 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+// If no explicit target flags are present, this defaults to the default
+// LLVM target.
+//
+// Works to handle the most common flags that modify the expected target such as
+// well as direct target flags.
```
Extra word?
```suggestion
// Works to handle the most common flags that modify the expected target as
// well as direct target flags.
```

jonmeow commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -157,4 +351,172 @@ auto ClangRunner::Run(llvm::ArrayRef<llvm::StringRef> args) -> bool {
+
+  // We use the object's address to avoid collisions between two instances,
+  // whether from two processes or two instances in this process. This is a bit
+  // of a hack perfect, but should be fine until we implement caching at which
```
What's a "hack perfect"?

chandlerc replied:
No idea, a typo.

jonmeow replied:
```suggestion
  // of a hack, but should be fine until we implement caching at which
```

jonmeow commented on toolchain/driver/clang_runner.h:
```diff
@@ -62,7 +88,37 @@ class ClangRunner : ToolRunnerBase {
+                        const std::filesystem::path& objs_path,
+                        llvm::StringRef out_path) -> void;
+
+  // A temp directory for this instance, will be cleaned up on destruction.
```
```suggestion
  // If non-empty, a temp directory for this instance which will be cleaned up on destruction. Set by `MakeTmpDir`.
```

jonmeow commented on bazel/llvm_project/0004_Introduce_filegroups_for_compiler_rt_builtins_runtime.patch:
```diff
@@ -0,0 +1,193 @@
+         "lib/fuzzer/Fuzzer*.def",
+     ]),
+     copts = [
+-      # Not using no-sanitize=address per https://github.com/google/sanitizers/wiki/AddressSanitizerContainerOverflow#false-positives
+-      "-fno-sanitize=memory,thread,undefined",
+-      "-fsanitize-coverage=0",
++        # Not using no-sanitize=address per https://github.com/google/sanitizers/wiki/AddressSanitizerContainerOverflow#false-positives
++        "-fno-sanitize=memory,thread,undefined",
++        "-fsanitize-coverage=0",
```
Sent you #5354 to try avoiding a patch of the patch

chandlerc replied:
Done.

jonmeow commented on toolchain/install/BUILD:
```diff
@@ -148,6 +148,16 @@ filegroup(
+# `builtins` directory.
+filegroup(
+    name = "clang_builtins_runtimes",
+    srcs = [
+        runtime
+        for runtime in CRT_FILES.values() + BUILTINS_FILEGROUPS.values()
+    ],
```
```suggestion
    srcs = CRT_FILES.values() + BUILTINS_FILEGROUPS.values(),
```

chandlerc replied:
Done.

jonmeow commented on bazel/llvm_project/0004_Introduce_filegroups_for_compiler_rt_builtins_runtime.patch:
```diff
@@ -0,0 +1,193 @@
++            "lib/builtins/*.h",
++            "lib/builtins/*.inc",
++        ],
++        exclude = BUILTINS_CRTBEGIN_SRCS + BUILTINS_CRTEND_SRCS + BUILTINS_TF_EXCLUDES + BUILTINS_TF_SRCS_PATTERNS + BUILTINS_MACOS_ATOMIC_SRCS_PATTERNS,
```
Minor, and don't bother if you don't want to regenerate the patch, but here and line 108, you might be able to add parens in order to wrap the additions (but I don't know how buildifier would react). i.e.:

```
exclude = (
  x +
  y +
  z
),
```

chandlerc replied:
Done since I was regenerating anyways.

jonmeow reviewed (commented):
The filegroup change looks great!

chandlerc reviewed (commented):
PTAL!

jonmeow commented on toolchain/install/BUILD:
```diff
@@ -193,7 +205,7 @@ make_install_filegroups(
 py_test(
     name = "llvm_symlinks_test",
-    size = "small",
+    size = "medium",
```
If this is taking too long to remain small, perhaps it could be sharded to try to keep `bazel test //...` fast?

Note, not something I think needs to be done here.

chandlerc replied:
This is no longer needed with the new version, removed.

chandlerc commented:
PTAL -- I think this is now ready for a fresh look.

Major changes since previous reviews:
- Switched to the new filesystem library everywhere I could.
- Adds a subcommand to directly build the runtimes, including to a designated destination directory.
- Supports passing an explicit pre-built dir to `clang` subcommand, and prebuilt resource-dir at the API level.

Combined this makes it pretty reasonable both to test out the new functionality of these being on-demand, and to manually rig up working runtimes that don't have to be rebuilt each time. This also shows a path for more build-system-integrated runtimes management I hope. Still lots of TODOs around this though to wire up a more comprehensive runtimes abstraction and connect that to a caching layer. But I wanted to try and get this in before it grows even larger.

Only the last commit should be reviewed here, this now depends on https://github.com/carbon-language/carbon-lang/pull/5968 which factors out some missing functions in the filesystem library.

zygoloid commented on toolchain/base/runtime_sources.bzl:
```diff
@@ -0,0 +1,157 @@
+- CRT: The C language runtimes not provided by the C standard library, currently
+  just infrastructure for global initialization and teardown.
+- Builtins: The compiler builtins library mirroring `libgcc` that provides
+  function definitions for operations not reliably available in hardware bet
```
```suggestion
  function definitions for operations not reliably available in hardware but
```

zygoloid commented on toolchain/driver/build_runtimes_subcommand.cpp:
```diff
@@ -0,0 +1,86 @@
+Directly running this command is not necessary as Carbon will work to build the
+runtimes needed on-demand when linking, but directly building them can aid in
+debugging issues or allow building them with a specific set of code generation
+flags and then reusing them in links explicitly.
```
Some minor copy-editing:
```suggestion
This subcommand builds Carbon's runtime libraries for a particular code
generation target, either in their default location or a specified one.

Running this command directly is not necessary as Carbon will build and cache
runtimes as needed when linking, but building them directly can aid in
debugging issues or allow them to be prebuilt, possibly with customized code
generation flags, and used explicitly when linking.
```

zygoloid commented on toolchain/driver/build_runtimes_subcommand.cpp:
```diff
@@ -0,0 +1,86 @@
+
+  // Don't run Clang when fuzzing, it is known to not be reliable under fuzzing
+  // due to many unfixed issues.
+  if (!DisableFuzzingExternalLibraries(driver_env, "clang")) {
```
(Out of scope for this PR, but I found this call a bit confusing. The return value seems backwards -- false if the command should be disabled, true if not. Maybe a different name would help but nothing short seems to clearly indicate what this does. `CheckNotFuzzingExternalLibrary` is the best I've found so far, but I don't think it's great.)

chandlerc replied:
(sending a PR to rename to `TestAndDiagnoseIfFuzzingExternalLibraries` and have it return true when fuzzing and diagnosed)

zygoloid commented on toolchain/driver/build_runtimes_subcommand.cpp:
```diff
@@ -0,0 +1,86 @@
+
+  auto tmp_result = Filesystem::MakeTmpDir();
+  if (!tmp_result.ok()) {
+    // FIXME: diagnose
```
I note this is a FIXME not a TODO, did you intend to do this in this PR?

chandlerc replied:
I intended to do this in this PR, thanks for flagging. Now done.

dwblaikie commented on toolchain/driver/build_runtimes_subcommand.cpp:
```diff
@@ -0,0 +1,86 @@
+
+  auto tmp_result = Filesystem::MakeTmpDir();
+  if (!tmp_result.ok()) {
+    // FIXME: diagnose
```
```suggestion
    // TODO: diagnose
```

chandlerc replied:
I intended to do this and just forgot. Done now.

dwblaikie commented on toolchain/driver/clang_runner_test.cpp:
```diff
@@ -195,5 +175,123 @@ TEST(ClangRunnerTest, CompileMultipleFiles) {
+  }
+
+  // Across all targets, check that the builtins archive exists, and contains a
+  // relevant symbol by running the `llvm-nm` tool over it. Using `nm` rather
+  // than directly inspecting the objects is a bit awkward, but lets us easily
+  // ignore the wrapping in an archive file.
```
huh, LLVM doesn't have a terse enough API for this? The wrapping is handled in the llvm-nm tool itself, rather than in a reusable LLVM library? That's unfortunate.

chandlerc replied:
I think the wrapping is handled in a reusable library, but requires a *lot* of boilerplate code to step through and query for specific symbols. =/ At least, I didn't see any convenient API to use in a test like this.

dwblaikie reviewed (approved):
C++ parts look good to me - and sounds like the bazel stuff had already had a pass.

zygoloid commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +51,214 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+// Works to handle the most common flags that modify the expected target as
+// well as direct target flags.
+//
+// Note: this has known fidelity issues if the args include separate-value flags
+// (`--flag value` style as opposed to `--flag=value`) where the value might
+// match the spelling of one of the target flags. For example, args that include
+// an output file spelled `-m32` (so `-o` followed by `-m32`) will be
+// misinterpreted by considering the value to itself be a flag. Addressing this
+// would add substantial complexity, including likely parsing the entire args
+// twice with the Clang driver. Instead, our current plan is to document this
+// limitation and encourage the use of flags with joined values
+// (`--flag=value`).
```
We already have logic to convert a sequence of clang driver arguments correctly to a set of clang frontend options in `Carbon::BuildClangInvocation`. I think we'll probably end up with all of our clang-invoking actions calling that anyway, to prepare to parse any imported C++ code, so perhaps we could reuse it here to obtain the target information reliably? Or would that fail if the resource directory for the target doesn't exist yet?

chandlerc replied:
It's not that the driver fails, but it computes the wrong command. I was trying to avoid handling the args twice.

Not sure we can reuse the `BuildClangInvocation`, because these args may not build a `ClangInvocation` as they're often just link commands. Might be flags that result in errors or warnings when used with a compilation too. =/

I can switch to running the args through the driver twice if we're really worried about this, but might ask for that in a follow-up?

zygoloid replied:
Yeah, definitely don't block this PR on this.

zygoloid commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +51,214 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+  auto result =
+      BuildTargetResourceDir(target, resource_dir_path, tmp_dir.abs_path());
+  if (!result.ok()) {
+    // FIXME: Need to diagnose this.
```
Did you mean to address this FIXME before merging?

chandlerc replied:
Yes, done. And thanks for flagging.

This one in particular required massaging the top level API a bit and subsequent changes to all the callers. =/

zygoloid commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +51,214 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+
+  llvm::Triple target_triple(target);
+  CARBON_CHECK(!target_triple.isOSWindows(),
+               "TODO: Windows runtimes are untested and not yet supported.");
```
Would it make sense to return an `Error` here instead of CHECK-failing?

chandlerc replied:
No preference, switched.

zygoloid commented on toolchain/driver/clang_runner.cpp:
```diff
@@ -42,24 +51,214 @@ auto clang_main(int Argc, char** Argv, const llvm::ToolContext& ToolContext)
+               "TODO: Windows runtimes are untested and not yet supported.");
+
+  // Create a directory with a known path that we can use for object files. We
+  // use an explicit `_objs` subdirectory for clarity when debugging.
+  CARBON_ASSIGN_OR_RETURN(Filesystem::Dir tmp_dir,
+                          Filesystem::Cwd().OpenDir(tmp_path));
+  CARBON_ASSIGN_OR_RETURN(Filesystem::Dir objs_dir,
+                          tmp_dir.OpenDir("_objs", Filesystem::OpenAlways));
+  std::filesystem::path objs_path = tmp_path / "_objs";
```
Do we need to create the CRT files in the "wrong" place and then moving them to the intended location? Clang already does that dance itself (changing the filename not the directory), so I think this will result in Clang creating the file in the wrong directory with the wrong filename, then Clang renaming it from the wrong filename to the right filename in the wrong directory, then us renaming it from the wrong directory to the right directory.

If this is important, using the same name `_objs` that bazel uses seems potentially confusing; maybe something like `build` (or just dropping the leading `_`) would be clearer that this isn't being built by bazel.

chandlerc replied:
It's important for the builtins archive -- we don't ever want the objects there.

I'd done the rename here mostly for consistency with the archive, and to consistently not depend on having an absolute path for the resource directory.

I can switch back if you prefer, but it will mean that we *must* be able to spell the resource directory path as part of the Clang argument list, rather than only requiring that of the temporary path. I guess we can cross that bridge if we ever end up caring.

I've just removed the `_objs` path component and passed the temp directory on through. Not sure the extra directory component was adding any value here.

zygoloid commented on toolchain/driver/clang_runner.h:
```diff
@@ -61,12 +97,47 @@ class ClangRunner : ToolRunnerBase {
+  // `BuildTargetResourceDir` to construct the desired runtimes directory first,
+  // and then set it here for subsequent use. Note that setting this disables
+  // any on-demand building of runtimes.
+  auto SetTargetRuntimesDir(const std::filesystem::path& target_runtimes_path)
+      -> void;
```
Is this left over from a previous iteration? Looks like it's not defined or used.

chandlerc replied:
Oops, yes. Removed.

zygoloid commented on toolchain/driver/clang_runner.h:
```diff
@@ -61,12 +97,47 @@ class ClangRunner : ToolRunnerBase {
  private:
-  llvm::StringRef target_;
-  llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem> fs_;
+  auto RunInternal(llvm::ArrayRef<llvm::StringRef> args, llvm::StringRef target,
```
Maybe move the function comment from the `.cpp` file to here.

chandlerc replied:
Done.

chandlerc committed:
Apply suggestions from code review

Co-authored-by: Richard Smith <richard@metafoo.co.uk>

chandlerc committed:
review fixes

chandlerc reviewed (commented):
Thanks, PTAL!

chandlerc committed:
fix from rebase

chandlerc commented:
Thanks all, merging!
